### PR TITLE
Remove usages of deprecated (in Java 9 and above) `Class#newInstance`

### PR DESCRIPTION
**Pull request by basil**
### Problem

The documentation for [`Class#newInstance`](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#newInstance--) includes the following warning:

> Note that this method propagates any exception thrown by the nullary constructor, including a checked exception. Use of this method effectively bypasses the compile-time exception checking that would otherwise be performed by the compiler. The [`Constructor#newInstance`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Constructor.html#newInstance-java.lang.Object...-) method avoids this problem by wrapping any exception thrown by the constructor in a (checked) [`InvocationTargetException`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html).

For this reason, the `Class#newInstance` method has been deprecated in Java 9 and above.

### Solution

We should always prefer `myClass.getConstructor().newInstance()` to calling `myClass.newInstance()` directly. The latter sequence of calls is inferred to be able to throw the additional exception types [`InvocationTargetException`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html) and [`NoSuchMethodException`](https://docs.oracle.com/javase/8/docs/api/java/lang/NoSuchMethodException.html). Both of these exception types are subclasses of [`ReflectiveOperationException`](https://docs.oracle.com/javase/8/docs/api/java/lang/ReflectiveOperationException.html). This PR updates our usages accordingly.

### Bonus

In addition, I noticed this codebase was using a mixture of `Error` and `AssertionError` for "impossible" reflection errors. [Error Prone](https://errorprone.info/bugpattern/RethrowReflectiveOperationExceptionAsLinkageError) recommends against this in favor of `LinkageError` for rethrowing `ReflectiveOperationException` exceptions as unchecked. While the reasoning isn't explicitly stated, presumably it is because `LinkageError` is more specific. I've updated our usage to match this recommendation.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/cli/Connection.java
Comment (jglick) [inline core/src/main/java/hudson/cli/Connection.java:271]: Arguably should be `AssertionError` because we believe it should never happen.

File: core/src/main/java/hudson/lifecycle/Lifecycle.java
Comment (jglick) [inline core/src/main/java/hudson/lifecycle/Lifecycle.java:96]: _Why checked exceptions were a terrible idea: Exhibit A_

File: core/src/main/java/hudson/util/DescriptorList.java
Comment (jglick) [inline core/src/main/java/hudson/util/DescriptorList.java:215]: `AssertionError` was correct here. Read the above lines carefully.

**Review by jglick — APPROVED** [review]
Fine if it avoids a handful of deprecation warnings whenever we do switch the codebase to `-source 11`. The loss of checked exceptions from `newInstance()` calls was not actually a problem at all—if you are using reflection obviously you do not know or care what `throws` clause the constructor used.

File: core/src/main/java/hudson/util/DescriptorList.java
Comment (timja) [reply core/src/main/java/hudson/util/DescriptorList.java:215]: seems to be in the description?

> In addition, I noticed this codebase was using a mixture of Error and AssertionError for "impossible" reflection errors. Error Prone recommends against this in favor of LinkageError for rethrowing ReflectiveOperationException exceptions as unchecked. While the reasoning isn't explicitly stated, presumably it is because LinkageError is more specific. I've updated our usage to match this recommendation.

Comment (jglick) [reply core/src/main/java/hudson/util/DescriptorList.java:215]: `LinkageError` makes sense if you were actually trying to link something and just happened to be using reflection to do it rather than a static linkage, where the same underlying problem could be thrown in either case with different `Throwable` types: that the actual version of one class is binary incompatible with the expected version. That is not the case here. We expect a `ClassLoader` to be able to `loadClass` a name which we know it already loaded. If that fails, something is wrong with program logic, and it is _not_ a matter of version skew between components. So a `@SuppressWarnings("RethrowReflectiveOperationExceptionAsLinkageError")` would be a more appropriate correction here. Obviously it does not much matter either way.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/util/DescriptorList.java
Comment (timja) [reply core/src/main/java/hudson/util/DescriptorList.java:215]: I think we can just go with it, it makes the tool happy and it's not really a problem...

given you approved it I assume you don't feel too strongly on it

File: core/src/main/java/hudson/cli/Connection.java
Comment (basil) [reply core/src/main/java/hudson/cli/Connection.java:271]: Agreed.

File: core/src/main/java/hudson/util/DescriptorList.java
Comment (basil) [reply core/src/main/java/hudson/util/DescriptorList.java]: Agreed.

File: core/src/main/java/hudson/lifecycle/Lifecycle.java
Comment (basil) [reply core/src/main/java/hudson/lifecycle/Lifecycle.java:96]: Indeed. I was tempted to use [Cy Young](https://www.launchableinc.com/a-tiny-library-for-java-exceptions-cy-young). That's probably what Kohsuke would have done back in the day. 😊 But I decided to keep it simple.

Comment (oleg-nenashev) [reply core/src/main/java/hudson/lifecycle/Lifecycle.java:96]: This is probably a case when introducing a new Checked Exception throwing method AND deprecating the old one is reasonable. I do not intend to start a holywar here though :)

**Review by oleg-nenashev — APPROVED** [review]
No objections though I would have implemented it differently

File: core/src/main/java/hudson/lifecycle/Lifecycle.java
Comment (jglick) [reply core/src/main/java/hudson/lifecycle/Lifecycle.java:96]: In this case the appropriate code would really be more like

```java
try {
    ClassLoader cl = Jenkins.get().getPluginManager().uberClassLoader;
    instance = (Lifecycle)cl.loadClass(p).getDeclaredConstructor().newInstance();
} catch (Throwable t) {
    LOGGER.log(Level.WARNING, "Could not load requested ${hudson.lifecycle}", t);
    instance = new ExitLifecycle();
}
```

**Pull request merged**
